utils/man/energymon-power-poller.1
.TH "ENERGYMON\-POWER\-POLLER" "1" "2018-11-04" "energymon" "EnergyMon Utilities"
.SH "NAME"
.LP
energymon\-power\-poller \- print average power at regular intervals
.SH "SYNPOSIS"
.LP
\fBenergymon\-power\-poller\fP
[\fIOPTION\fP]...
.SH "DESCRIPTION"
.LP
Prints the average power at regular intervals.
Uses the default EnergyMon implementation.
.LP
If no additional options are specified, average power is printed to standard
output at the implementation's minimum update interval, and the program loops
until interrupted with \fBCTRL\-C\fP.
Note that using the default interval is not always desirable, as it may be too
fast and cause unnecessary overhead.
Also, computing power at the minimum update interval can result in noisy data.
.SH "OPTIONS"
.LP
.TP
\fB\-h\fP, \fB\-\-help\fP
Prints the help screen.
.TP
\fB\-c\fP, \fB\-\-count\fP=\fIN\fP
Stop after \fIN\fP reads.
.TP
\fB\-f\fP, \fB\-\-file\fP=\fIFILE\fP
Write the output to \fIFILE\fP.
.TP
\fB\-i\fP, \fB\-\-interval=\fP\fIUS\fP
The update interval in microseconds.
Must be greater than 0.
Note that if \fIUS\fP is too short for the EnergyMon implementation to get
updated energy readings, the average power will be reported as 0.
.TP
\fB\-s\fP, \fB\-\-summarize\fP
Print out a summary at completion.
.SH "EXAMPLES"
.TP
\fBenergymon\-power\-poller\fP
Measure average power at the implementation's update interval.
.TP
\fBenergymon\-power\-poller \-i 5000000 \-f power.log\fP
Measure average power every 5 seconds and write output to \fBpower.log\fP.
.TP
\fBenergymon\-power\-poller \-i 100000 \-c 10\fP
Measure average power every 100 milliseconds and stop after 10 readings.
.TP
\fBenergymon\-power\-poller \-i 50000 \-c 100 \-s\fP
Measure average power every 50 milliseconds, stop after 100 readings, and print
a summary.
.SH "BUGS"
.LP
Report bugs upstream at <https://github.com/energymon/energymon>
.SH "SEE ALSO"
.BR energymon\-file\-provider (1)

RELEASES.md
# Release Notes

## [Unreleased]
### Changed
 * More pedantic formatting in man pages

## [v0.2.4] - 2018-06-01
### Changed
 * WattsUp implementations:
   * Wait up to 5 seconds for a good data packet (or fail) before completing initialization
   * Purge data buffers during initialization
   * Poll device at shorter intervals to increase energy value accuracy
   * Better handling of incomplete data packets
 * osp-polling: Allow polling thread to be interrupted during sleep when finish is requested
 * energymon-power-poller: Account for true elapsed time to produce more accurate power values

### Fixed
 * wattsup-libftdi, wattsup-libusb: Fix empty reads by always setting baud rate and serial line properties
 * wattsup-dev: Use pselect to avoid timer expired errors
 * odroid, odroid-ioctl, osp-polling, zcu102: Minor correction to polling delay interval
 * msr: Possible (but unlikely) precision computation bug
 * osp, osp-polling: Fix status check during init that didn't always work


## [v0.2.3] - 2017-11-19
### Added
 * Implementation for Xilinx ZCU102 (Jason Miller)

### Fixed
 * Fixed [#40]: Time conversion bug in common function 'energymon_gettime_elapsed_us' (Jason Miller)


## [v0.2.2] - 2017-10-26
### Added
 * New capabilities to some utility applications:
   * energymon-cmd-profile: Accept commands without quotations; add help option
   * energymon-file-provider: Add help, count, interval, and no-rewind options
   * energymon-idle-power: Add help option
 * Man pages for utility applications
 * VERSION and SOVERSION properties on shared object libraries
 * Multiarch support (use GNU standard installation directories)

### Changed
 * Increased minimum CMake version from 2.8 to 2.8.5 to support GNUInstallDirs
 * Refactor this RELEASES.md file

### Removed
 * Removed private symbol exports in shared object libraries (prevents symbol name clashes)

### Fixed
 * Various bug fixes to utility applications


## [v0.2.1] - 2017-10-01
### Added
 * Support for msr-safe kernel module in msr implementation
 * New cray-pm implementations to support capabilities of Cray XC30 and XC40 systems
 * New wattsup-libusb and wattsup-libftdi implementations
 * AppVeyor test configuration for Windows continuous integration

### Changed
 * Faster refresh in osp-polling, odroid and odroid-ioctl implementations
 * Better multi-platform support for some implementations, especially with portable time and sleep abstractions
 * Downgraded required CMake version from 2.8.9 to 2.8
 * Documented FSE 2016 and Technical Report publication information
 * Various enhancements, optimizations, and documentation improvements

### Deprecated
 * Deprecated custom DEFAULT CMake option in favor of ENERGYMON_BUILD_DEFAULT

### Removed
 * Removed Android CMake toolchain file: cmake-toolchain/android.toolchain.cmake

### Fixed
 * Fixed [#37]: In osp implementation, force overflow in ODROID Smart Power at 1000 Wh to avoid device counter saturation at 8192 Wh
 * Fixed [#27]: More modular build process
 * Numerous other bugs

## v0.2.0 - 2016-10-12
### Added
 * Initial public release

[Unreleased]: https://github.com/energymon/energymon/compare/v0.2.4...HEAD
[v0.2.4]: https://github.com/energymon/energymon/compare/v0.2.3...v0.2.4
[v0.2.3]: https://github.com/energymon/energymon/compare/v0.2.2...v0.2.3
[v0.2.2]: https://github.com/energymon/energymon/compare/v0.2.1...v0.2.2
[v0.2.1]: https://github.com/energymon/energymon/compare/v0.2.0...v0.2.1
[#40]: https://github.com/energymon/energymon/issues/40
[#37]: https://github.com/energymon/energymon/issues/37
[#27]: https://github.com/energymon/energymon/issues/27


utils/energymon-power-poller.c
/**
 * Polls an energymon at the requested interval and prints the average power
 * over that interval.
 *
 * @author Connor Imes
 * @date 2016-08-30
 */
#define _GNU_SOURCE
#include <errno.h>
#include <float.h>
#include <getopt.h>
#include <inttypes.h>
#include <limits.h>
#include <math.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include "energymon-default.h"
#include "energymon-time-util.h"

static const int IGNORE_INTERRUPT = 0;

static volatile uint64_t running = 1;
static int count = 0;
static const char* filename = NULL;
static int summarize = 0;
static uint64_t interval = 0;

static const char short_options[] = "hc:f:i:s";
static const struct option long_options[] = {
  {"help",      no_argument,       NULL, 'h'},
  {"count",     required_argument, NULL, 'c'},
  {"file",      required_argument, NULL, 'f'},
  {"interval",  required_argument, NULL, 'i'},
  {"summarize", no_argument,       NULL, 's'},
  {0, 0, 0, 0}
};

static void print_usage(int exit_code) {
  fprintf(exit_code ? stderr : stdout,
          "Usage: energymon-power-poller [OPTION]...\n"
          "Options:\n"
          "  -h, --help               Print this message and exit\n"
          "  -c, --count=N            Stop after N reads\n"
          "  -f, --file=FILE          The output file\n"
          "  -i, --interval=US        The update interval in microseconds\n"
          "  -s, --summarize          Print out a summary at completion\n");
  exit(exit_code);
}

static void parse_args(int argc, char** argv) {
  int c;
  while ((c = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
    switch (c) {
      case 'h':
        print_usage(0);
        break;
      case 'c':
        count = 1;
        running = strtoull(optarg, NULL, 0);
        break;
      case 'f':
        filename = optarg;
        break;
      case 'i':
        interval = strtoull(optarg, NULL, 0);
        if (interval == 0) {
          fprintf(stderr, "Interval must be > 0\n");
          print_usage(1);
        }
        break;
      case 's':
        summarize = 1;
        break;
      case '?':
      default:
        print_usage(1);
        break;
    }
  }
}

static void shandle(int sig) {
  switch (sig) {
    case SIGTERM:
    case SIGINT:
#ifdef SIGQUIT
    case SIGQUIT:
#endif
#ifdef SIGKILL
    case SIGKILL:
#endif
#ifdef SIGHUP
    case SIGHUP:
#endif
      running = 0;
    default:
      break;
  }
}

int main(int argc, char** argv) {
  energymon em;
  uint64_t min_interval;
  uint64_t energy;
  uint64_t energy_last;
  uint64_t last_us;
  uint64_t exec_us;
  float power;
  uint64_t n = 0;
  float pmin = FLT_MAX;
  float pmax = FLT_MIN;
  float pavg = 0;
  float pavg_last = 0;
  float pstd = 0;
  FILE* fout = stdout;
  int ret = 0;

  // register the signal handler
  signal(SIGINT, shandle);

  parse_args(argc, argv);

  // initialize the energy monitor
  if (energymon_get_default(&em)) {
    perror("energymon_get_default");
    return 1;
  }
  if (em.finit(&em)) {
    perror("energymon:finit");
    return 1;
  }

  // get the update interval
  min_interval = em.finterval(&em);
  if (interval == 0) {
    interval = min_interval;
  } else if (interval < min_interval) {
    fprintf(stderr,
            "Requested interval is too short, minimum available: %"PRIu64"\n",
            min_interval);
    em.ffinish(&em);
    return 1;
  }

  // open the output file
  if (filename != NULL) {
    fout = fopen(filename, "w");
    if (fout == NULL) {
      perror(filename);
      em.ffinish(&em);
      return 1;
    }
  }

  // output at regular intervals
  energy_last = em.fread(&em);
  last_us = energymon_gettime_us();
  energymon_sleep_us(interval, &IGNORE_INTERRUPT);
  while (running) {
    if (count) {
      running--;
    }
    energy = em.fread(&em);
    exec_us = energymon_gettime_elapsed_us(&last_us);
    power = (energy - energy_last) / ((float) exec_us);
    if (fprintf(fout, "%f\n", power) < 0) {
      ret = 1;
      if (filename == NULL) {
        // should never happen writing to stdout
        perror("Writing");
      } else {
        perror(filename);
      }
      break;
    }
    fflush(fout);
    energy_last = energy;
    if (power > pmax) {
      pmax = power;
    }
    if (power < pmin) {
      pmin = power;
    }
    // get the running average and data to compute stddev
    pavg_last = pavg;
    n++;
    if (n == 1) {
      pavg = power;
    } else {
      pavg = pavg + ((power - pavg) / n);
    }
    pstd = pstd + (power - pavg) * (power - pavg_last);
    if (running) {
      energymon_sleep_us(interval, &IGNORE_INTERRUPT);
    }
  }

  if (summarize) {
    fprintf(fout, "Samples: %"PRIu64"\n", n);
    fprintf(fout, "Pavg: %f\n", pavg);
    fprintf(fout, "Pmax: %f\n", pmax);
    fprintf(fout, "Pmin: %f\n", pmin);
    fprintf(fout, "Pstdev: %f\n", n == 1 ? 0 : (sqrt(pstd / (n - 1))));
    fprintf(fout, "Joules: %f\n", n * pavg * (interval / 1000000.0f));
  }

  // cleanup
  if (filename != NULL && fclose(fout)) {
    perror(filename);
  }
  if (em.ffinish(&em)) {
    perror("em.ffinish");
  }

  return ret;
}
